cmd/microcloud/ask.go

```diff
@@ -589,17 +671,82 @@ func (c *CmdControl) askNetwork(sh *service.Handler, systems map[string]InitSyst
 		return nil
 	}
 
-	// Environments without OVN get a basic fan setup.
-	if sh.Services[types.MicroOVN] == nil {
-		return nil
-	}
-
 	// Get the list of networks from all peers.
 	infos := []mdns.ServerInfo{}
 	for _, system := range systems {
 		infos = append(infos, system.ServerInfo)
 	}
 
+	// Configure the Ceph networks.
+	lxdService := sh.Services[types.LXD].(*service.LXDService)
+	if lxdService == nil {
+		return fmt.Errorf("failed to get LXD service")
+	}
+
+	availableCephNetworkInterfaces, err := lxdService.GetCephInterfaces(context.Background(), bootstrap, infos)
+	if err != nil {
+		return err
+	}
+
+	if len(availableCephNetworkInterfaces) == 0 {
+		fmt.Println("No network interfaces found with IPs to set a dedicated Ceph network, skipping Ceph network setup")
+	} else {
+		if bootstrap {
+			microCloudInternalNetworkAddr := microCloudInternalSubnet.IP.Mask(microCloudInternalSubnet.Mask)
+			ones, _ := microCloudInternalSubnet.Mask.Size()
+			microCloudInternalNetworkAddrCIDR := fmt.Sprintf("%s/%d", microCloudInternalNetworkAddr.String(), ones)
+
+			publicCephSubnet, err := c.asker.AskString(fmt.Sprintf("What subnet (either IPv4 or IPv6 CIDR notation) would you like your Ceph public traffic on? [default: %s] ", microCloudInternalNetworkAddrCIDR), microCloudInternalNetworkAddrCIDR, validate.IsNetwork)
+			if err != nil {
+				return err
+			}
+
+			if publicCephSubnet != microCloudInternalNetworkAddrCIDR {
+				err = validateCephInterfacesForSubnet(lxdService, systems, availableCephNetworkInterfaces, publicCephSubnet)
+				if err != nil {
+					return err
+				}
+
+				bootstrapSystem := systems[sh.Name]
+				bootstrapSystem.MicroCephPublicNetworkSubnet = publicCephSubnet
+				systems[sh.Name] = bootstrapSystem
+			}
+
+			// If a user chose a network different from the default one,
+			// we make the default for the internal network the same as the first one.
+			// This avoid entering two times the same network in order to have a partially disaggregated setup.
+			if publicCephSubnet != microCloudInternalNetworkAddrCIDR {
+				microCloudInternalNetworkAddrCIDR = publicCephSubnet
+			}
+
+			internalCephSubnet, err := c.asker.AskString(fmt.Sprintf("What subnet (either IPv4 or IPv6 CIDR notation) would you like your Ceph internal traffic on? [default: %s] ", microCloudInternalNetworkAddrCIDR), microCloudInternalNetworkAddrCIDR, validate.IsNetwork)
+			if err != nil {
+				return err
+			}
+
+			if internalCephSubnet != microCloudInternalNetworkAddrCIDR {
+				err = validateCephInterfacesForSubnet(lxdService, systems, availableCephNetworkInterfaces, internalCephSubnet)
+				if err != nil {
+					return err
+				}
+
+				bootstrapSystem := systems[sh.Name]
+				bootstrapSystem.MicroCephInternalNetworkSubnet = internalCephSubnet
+				systems[sh.Name] = bootstrapSystem
+			}
+		} else {
+			err = validateAddingMemberWithCephNetwork(sh, lxdService, systems, availableCephNetworkInterfaces, microCloudInternalSubnet.String())
```

Comment:
Since we can be re-using an existing MicroCeph cluster now, we can't just rely on the unix socket anymore.

For example, lets say you want to set up MicroCloud with 3 systems.
- system1: all snaps are pristine
- system2: microceph has been bootstrapped, and is part of a 10-system cluster unrelated to microcloud
- system3: all snaps are pristine

In this case, we should end up with a 3-system MicroCloud/LXD/MicroOVN, and a 13 system MicroCeph.

But that means you need to connect to the existing (remote) MicroCeph deployment by specifying the address and mDNS auth secret of the remote system to the `Client` function.

You can grab the info of which systems are initialized from the `InitializedServices` map in `InitSystem`. You can look at this block to see how we handle something like this for issuing join tokens remotely if one of our systems is already clustered.
https://github.com/canonical/microcloud/blob/main/cmd/microcloud/main_init.go#L379-L435

Reply:
To clarify, I'm talking about fetching the cluster config from MicroCeph.

Reply:
In your case, since this endpoint already exists over the network in MicroCeph, you don't have to create a manual endpoint like we did for the join tokens, it would suffice to just specify the address & secret of system2 to the Client function.

Reply:
Understood! What if during the add, multiple systems are part of different MicroCeph clusters? If this is even possible, here is the validation process I had in mind:
```go
func validateAddingMemberWithCephNetwork(sh *service.Handler, lxd *service.LXDService, systems map[string]InitSystem, cephInterfaces map[string][]service.CephDedicatedInterface, existingMicroCloudInternalSubnet string) error {
	// In case of new machines added to the cluster that will be part of an existing Ceph cluster,
	// and if a dedicated Ceph cluster network has been configured, we need to ensure that the new machines
	// have at least one network interface with an IP within the existing Ceph network subnet.
	microCephService := sh.Services[types.MicroCeph].(*service.CephService)
	if microCephService != nil {
		// Separate the systems that have been initialized with the MicroCeph service from the ones that haven't.
		// There might be multiple 'remote' MicroCeph clusters to validate against because among the systems to be added,
		// there might be some that have already been initialized with the MicroCeph service. See the example below:
		//
		// 			    MicroCeph cluster A
		// 		            +------------------------------+
		// 			    |                              |
		// 			    |                              |
		// 			    |                              |
		// 			    |                +-------+     |
		// 		            |        +-------+       |     |
		// 			    |        |       |       |     |
		// 			    |        |       +-------+     |
		// 			    |        |       +-------+     |
		// 			    |        | +-----+       |     |
		// 			    |        | |     |       |     |
		// Existing MicroCloud      |    +---+-++    +-------+     |
		// +---------+              |    | ADD  |    +-------+     |
		// |         |    +-------------->      +----+       |     |
		// | +-----+ |    |         |    +------+    |       |     |
		// | |     | |    |         |                +-------+     |
		// | |     | |    |         +------------------------------+
		// | +-----+ |    |
		// |         |    |          Pristine node
		// | +-----+ |    |          +------+
		// | |  SH +-----------------> ADD  |
		// | |     | |    |          |      |
		// | +-----+ |    |          +------+
		// |         |    |
		// | +-----+ |    |         MicroCeph cluster B
		// | |     | |    |         +--------------------------------+
		// | |     | |    |         |                                |
		// | +-----+ |    |         |    +------+          +------+  |
		// |         |    +--------------> ADD  +----------+      |  |
		// +---------+              |    |      |          |      |  |
		// 	                    |    +----+-+          +------+  |
		// 	                    |         |                      |
		// 	                    |         |            +------+  |
		// 	                    |         +------------+      |  |
		// 	                    |                      |      |  |
		// 	                    |                      +------+  |
		// 	                    |                                |
		// 	                    +--------------------------------+
		//
		// * In such a scenario, the 'initialized' nodes to be added (with their remote Ceph cluster configuration),
		//   will be validated one by one against the existing MicroCloud with a Ceph deployment.
		// * The 'uninitialized' nodes will be validated all together against the existing MicroCloud with a Ceph deployment.

		initializedMicroCephSystems := make(map[string]InitSystem)
		uninitializedMicroCephSystems := make(map[string]InitSystem)
		for peer, system := range systems {
			if system.InitializedServices[types.MicroCeph][peer] != "" {
				initializedMicroCephSystems[peer] = system
			} else {
				uninitializedMicroCephSystems[peer] = system
			}
		}

		checkCephConfigs := func(systems map[string]InitSystem, initialized bool) error {
			var targetAddress string
			var targetAuthSecret string
			for peer, system := range systems {
				if initialized {
					targetAddress = system.ServerInfo.Address
					targetAuthSecret = system.ServerInfo.AuthSecret
				}

				cephConfigs, err := microCephService.ClusterConfig(context.Background(), targetAddress, targetAuthSecret)
				if err != nil {
					return err
				}

				// Read the Ceph network configuration from the existing Ceph cluster
				existingCephInternalSubnet := ""
				existingCephPublicSubnet := ""
				for key, value := range cephConfigs {
					if key == "cluster_network" && value != "" {
						// Sometimes, the default cluster_network value in the Ceph configuration
						// is not a network range but a regular IP address. We need to extract the network range.
						_, valueNet, err := net.ParseCIDR(value)
						if err != nil {
							return fmt.Errorf("failed to parse the Ceph cluster network configuration from the existing Ceph cluster: %v", err)
						}

						if valueNet.String() != existingMicroCloudInternalSubnet {
							existingCephInternalSubnet = valueNet.String()
						}
					}

					if key == "public_network" && value != "" {
						_, valueNet, err := net.ParseCIDR(value)
						if err != nil {
							return fmt.Errorf("failed to parse the Ceph public network configuration from the existing Ceph cluster: %v", err)
						}

						if valueNet.String() != existingMicroCloudInternalSubnet {
							existingCephPublicSubnet = valueNet.String()
						}
					}
				}

				if initialized {
					if existingCephInternalSubnet != "" {
						err = validateCephInterfacesForSubnet(lxd, map[string]InitSystem{peer: system}, cephInterfaces, existingCephInternalSubnet)
						if err != nil {
							return err
						}
					}

					if existingCephPublicSubnet != "" {
						err = validateCephInterfacesForSubnet(lxd, map[string]InitSystem{peer: system}, cephInterfaces, existingCephPublicSubnet)
						if err != nil {
							return err
						}
					}
				} else {
					if existingCephInternalSubnet != "" {
						err = validateCephInterfacesForSubnet(lxd, systems, cephInterfaces, existingCephInternalSubnet)
						if err != nil {
							return err
						}
					}

					if existingCephPublicSubnet != "" {
						err = validateCephInterfacesForSubnet(lxd, systems, cephInterfaces, existingCephPublicSubnet)
						if err != nil {
							return err
						}
					}

					// The uninitialized nodes can be validated all together as there is no remote Ceph cluster configuration to validate against,
					// so we can break the loop here.
					break
				}
			}

			return nil
		}

		err := checkCephConfigs(uninitializedMicroCephSystems, false)
		if err != nil {
			return err
		}

		if len(initializedMicroCephSystems) > 0 {
			err = checkCephConfigs(initializedMicroCephSystems, true)
			if err != nil {
				return err
			}
		}

		return nil
	}

	return nil
}
```

Reply:
> @masnax Understood! What if during the add, multiple systems are part of different MicroCeph clusters? 

We can't reconcile 2 different clusters for the same service, so the supported configurations are strictly like this:

* All systems are unclustered before running `microcloud init`
* A subset (so 1 or more, up to all) of the systems being added to the MicroCloud are part of the same pre-existing cluster for MicroCeph or MicroOVN
  * The subset contains external systems we don't care about in MicroCloud. These can be ignored.

So the key here is that there will only ever be 0 or 1 external clusters for each service, but the participants can vary. Additionally, we only need to support this for `microcloud init`, not `microcloud add` since if running `microcloud add` the system we are interacting with must necessarily be a cluster member already.

So for this I think the logical flow should look something like this: 

* If we are bootstrapping: 
  * We check if there are any initialized systems for MicroCeph. If there is then we save the first one we encounter to issue remote requests to. We can assume if there is more than 1 that they are part of the same cluster.
  * We target the remote MicroCeph and fetch its cluster config. If it has not set up cluster/public networks, then we ask the questions. If it has set up cluster/public networks, then we try to see if the systems joining that cluster have the same subnets available.
* If we are not bootstrapping (so adding):
  * We can assume `systems[sh.Name]` (ourself) is clustered.
  * We target the local MicroCeph and fetch its cluster config. If it has not set up cluster/public networks, then we ask the questions. If it has set up cluster/public networks, then we try to see if the systems joining that cluster have the same subnets available.


I noticed currently you don't ask any questions in the `add` case. It might just be my misunderstanding, but what is MicroCeph's default behaviour if cluster/public networks have not been manually configured? Does it pick some defaults? If so then we probably don't need to ask questions in those cases (both for bootstrap with a pre-existing cluster & all add cases). But if the configs are unset then we should probably ask questions to set them.


Hope that's clear, let me know if it's confusing :)

Reply:
Very clear! thanks!

Reply:
To answer your question regarding the `add` case not having any questions:

The logic is to only validate that the to-be-added systems have network interfaces falling within the same subnet as the one configured in the cluster. If no ceph network have been setup in the cluster, this validation process is skipped.

The default behavior of MicroCeph if no cluster/public network have been configured is to use the MicroCloud internal network.

Reply:
I see, so if I `microceph cluster bootstrap`, it will automatically set some value for `cluster_network` and `public_network`?

Reply:
@gabrielmougard looks like the tests are not passing, something to do with the subnet being `nil`.

Also, could you please confirm the question I asked In my [previous comment](https://github.com/canonical/microcloud/pull/274#discussion_r1612038549)?